File: src/node/tables/os2.cc
#include <nan.h>
#include <ft2build.h>
#include "../freetype.h"
#include FT_TRUETYPE_TABLES_H

NAN_METHOD(FreeType::GetOS2Table) {
    NanScope();
    FreeType* ff = ObjectWrap::Unwrap<FreeType>(args.This());
    TT_OS2* os2 = (TT_OS2*) FT_Get_Sfnt_Table(ff->face, ft_sfnt_os2);
    if(os2 == NULL) {
        NanReturnNull();
    }

//   typedef struct  TT_OS2_
//    {
//      FT_UShort  version;                /* 0x0001 - more or 0xFFFF */
//      FT_Short   xAvgCharWidth;
//      FT_UShort  usWeightClass;
//      FT_UShort  usWidthClass;
//      FT_Short   fsType;
//      FT_Short   ySubscriptXSize;
//      FT_Short   ySubscriptYSize;
//      FT_Short   ySubscriptXOffset;
//      FT_Short   ySubscriptYOffset;
//      FT_Short   ySuperscriptXSize;
//      FT_Short   ySuperscriptYSize;
//      FT_Short   ySuperscriptXOffset;
//      FT_Short   ySuperscriptYOffset;
//      FT_Short   yStrikeoutSize;
//      FT_Short   yStrikeoutPosition;
//      FT_Short   sFamilyClass;
//
//      FT_Byte    panose[10];
//
//      FT_ULong   ulUnicodeRange1;        /* Bits 0-31   */
//      FT_ULong   ulUnicodeRange2;        /* Bits 32-63  */
//      FT_ULong   ulUnicodeRange3;        /* Bits 64-95  */
//      FT_ULong   ulUnicodeRange4;        /* Bits 96-127 */
//
//      FT_Char    achVendID[4];
//
//      FT_UShort  fsSelection;
//      FT_UShort  usFirstCharIndex;
//      FT_UShort  usLastCharIndex;
//      FT_Short   sTypoAscender;
//      FT_Short   sTypoDescender;
//      FT_Short   sTypoLineGap;
//      FT_UShort  usWinAscent;
//      FT_UShort  usWinDescent;
//
//      /* only version 1 and higher: */
//
//      FT_ULong   ulCodePageRange1;       /* Bits 0-31   */
//      FT_ULong   ulCodePageRange2;       /* Bits 32-63  */
//
//      /* only version 2 and higher: */
//
//      FT_Short   sxHeight;
//      FT_Short   sCapHeight;
//      FT_UShort  usDefaultChar;
//      FT_UShort  usBreakChar;
//      FT_UShort  usMaxContext;
//
//      /* only version 5 and higher: */
//
//      FT_UShort  usLowerOpticalPointSize;       /* in twips (1/20th points) */
//      FT_UShort  usUpperOpticalPointSize;       /* in twips (1/20th points) */
//
//    } TT_OS2;

    Local<Object> obj = Object::New();

    OBJ_SET_INTEGER("version", os2->version);
    OBJ_SET_INTEGER("xAvgCharWidth", os2->xAvgCharWidth);
    OBJ_SET_INTEGER("usWeightClass", os2->usWeightClass);
    OBJ_SET_INTEGER("usWidthClass", os2->usWidthClass);
    OBJ_SET_INTEGER("fsType", os2->fsType);
    OBJ_SET_INTEGER("ySubscriptXSize", os2->ySubscriptXSize);
    OBJ_SET_INTEGER("ySubscriptYSize", os2->ySubscriptYSize);
    OBJ_SET_INTEGER("ySubscriptXOffset", os2->ySubscriptXOffset);
    OBJ_SET_INTEGER("ySubscriptYOffset", os2->ySubscriptYOffset);
    OBJ_SET_INTEGER("ySuperscriptXSize", os2->ySuperscriptXSize);
    OBJ_SET_INTEGER("ySuperscriptYSize", os2->ySuperscriptYSize);
    OBJ_SET_INTEGER("ySuperscriptXOffset", os2->ySuperscriptXOffset);
    OBJ_SET_INTEGER("ySuperscriptYOffset", os2->ySuperscriptYOffset);
    OBJ_SET_INTEGER("yStrikeoutSize", os2->yStrikeoutSize);
    OBJ_SET_INTEGER("yStrikeoutPosition", os2->yStrikeoutPosition);
    OBJ_SET_INTEGER("sFamilyClass", os2->sFamilyClass);

    OBJ_SET_ARRAY("panose", os2->panose, 10);

    OBJ_SET_INTEGER("ulUnicodeRange1", os2->ulUnicodeRange1);        /* Bits 0-31   */
    OBJ_SET_INTEGER("ulUnicodeRange2", os2->ulUnicodeRange2);        /* Bits 32-63  */
    OBJ_SET_INTEGER("ulUnicodeRange3", os2->ulUnicodeRange3);        /* Bits 64-95  */
    OBJ_SET_INTEGER("ulUnicodeRange4", os2->ulUnicodeRange4);        /* Bits 96-127 */
    //

    OBJ_SET_ARRAY("achVendID", os2->achVendID, 4);

    OBJ_SET_INTEGER("fsSelection", os2->fsSelection);
    OBJ_SET_INTEGER("usFirstCharIndex", os2->usFirstCharIndex);
    OBJ_SET_INTEGER("usLastCharIndex", os2->usLastCharIndex);
    OBJ_SET_INTEGER("sTypoAscender", os2->sTypoAscender);
    OBJ_SET_INTEGER("sTypoDescender", os2->sTypoDescender);
    OBJ_SET_INTEGER("sTypoLineGap", os2->sTypoLineGap);
    OBJ_SET_INTEGER("usWinAscent", os2->usWinAscent);
    OBJ_SET_INTEGER("usWinDescent", os2->usWinDescent);
    //
    //      /* only version 1 and higher: */
    //
    OBJ_SET_INTEGER("ulCodePageRange1", os2->ulCodePageRange1);       /* Bits 0-31   */
    OBJ_SET_INTEGER("ulCodePageRange2", os2->ulCodePageRange2);       /* Bits 32-63  */
    //
    //      /* only version 2 and higher: */
    //
    OBJ_SET_INTEGER("sxHeight", os2->sxHeight);
    OBJ_SET_INTEGER("sCapHeight", os2->sCapHeight);
    OBJ_SET_INTEGER("usDefaultChar", os2->usDefaultChar);
    OBJ_SET_INTEGER("usBreakChar", os2->usBreakChar);
    OBJ_SET_INTEGER("usMaxContext", os2->usMaxContext);
    //
    //      /* only version 5 and higher: */
    //
    OBJ_SET_INTEGER("usLowerOpticalPointSize", os2->usLowerOpticalPointSize);       /* in twips (1/20th points) */
    OBJ_SET_INTEGER("usUpperOpticalPointSize", os2->usUpperOpticalPointSize);       /* in twips (1/20th points) */

    NanReturnValue(obj);
}

File: lib/generate/tables/os2.js
'use strict';

// See documentation here: http://www.microsoft.com/typography/otspec/os2.htm

var _ = require('underscore');
var ByteBuffer = require('../byte_buffer.js');

//get first glyph unicode
function getFirstCharIndex(font) {
    var minGlyph = _.min(font.glyphs, function (glyph) {
        return (glyph.unicode === 0) ? 0xFFFF : (glyph.unicode || 0);
    });
    if (minGlyph) {
        return minGlyph.unicode > 0xFFFF ? 0xFFFF : (minGlyph.unicode || 0);
    } else {
        return 0xFFFF;
    }
}

//get last glyph unicode
function getLastCharIndex(font) {
    var maxGlyph = _.max(font.glyphs, 'unicode');
    if (maxGlyph) {
        return maxGlyph.unicode > 0xFFFF ? 0xFFFF : (maxGlyph.unicode || 0);
    } else {
        return 0xFFFF;
    }
}

function createOS2Table(font) {

    var buf = ByteBuffer.prototype.create(86);
    var os2 = font.tables.os2;

    buf.writeUint16(1); //version
    buf.writeInt16(os2.avgWidth); // xAvgCharWidth
    buf.writeUint16(os2.weightClass); // usWeightClass
    buf.writeUint16(os2.widthClass); // usWidthClass
    buf.writeInt16(os2.fsType); // fsType
    buf.writeInt16(os2.ySubscriptXSize); // ySubscriptXSize
    buf.writeInt16(os2.ySubscriptYSize); //ySubscriptYSize
    buf.writeInt16(os2.ySubscriptXOffset); // ySubscriptXOffset
    buf.writeInt16(os2.ySubscriptYOffset); // ySubscriptYOffset
    buf.writeInt16(os2.ySuperscriptXSize); // ySuperscriptXSize
    buf.writeInt16(os2.ySuperscriptYSize); // ySuperscriptYSize
    buf.writeInt16(os2.ySuperscriptXOffset); // ySuperscriptXOffset
    buf.writeInt16(os2.ySuperscriptYOffset); // ySuperscriptYOffset
    buf.writeInt16(os2.yStrikeoutSize); // yStrikeoutSize
    buf.writeInt16(os2.yStrikeoutPosition); // yStrikeoutPosition
    buf.writeInt16(os2.familyClass); // sFamilyClass
    buf.writeUint8(os2.panose[0]); // panose.bFamilyType
    buf.writeUint8(os2.panose[1]); // panose.bSerifStyle
    buf.writeUint8(os2.panose[2]); // panose.bWeight
    buf.writeUint8(os2.panose[3]); // panose.bProportion
    buf.writeUint8(os2.panose[4]); // panose.bContrast
    buf.writeUint8(os2.panose[5]); // panose.bStrokeVariation
    buf.writeUint8(os2.panose[6]); // panose.bArmStyle
    buf.writeUint8(os2.panose[7]); // panose.bLetterform
    buf.writeUint8(os2.panose[8]); // panose.bMidline
    buf.writeUint8(os2.panose[9]); // panose.bXHeight
    buf.writeUint32(os2.ulUnicodeRange1); // ulUnicodeRange1
    buf.writeUint32(os2.ulUnicodeRange2); // ulUnicodeRange2
    buf.writeUint32(os2.ulUnicodeRange3); // ulUnicodeRange3
    buf.writeUint32(os2.ulUnicodeRange4); // ulUnicodeRange4
    buf.writeUint8(os2.achVendID[0]); //achVendID
    buf.writeUint8(os2.achVendID[1]);
    buf.writeUint8(os2.achVendID[2]);
    buf.writeUint8(os2.achVendID[3]);
    buf.writeUint16(os2.fsSelection); // fsSelection
    buf.writeUint16(getFirstCharIndex(font)); // usFirstCharIndex
    buf.writeUint16(getLastCharIndex(font)); // usLastCharIndex
    buf.writeInt16(os2.sTypoAscender); // sTypoAscender
    buf.writeInt16(os2.sTypoDescender); // sTypoDescender
    buf.writeInt16(os2.sTypoLineGap); // lineGap
    // Enlarge win acscent/descent to avoid clipping
    buf.writeInt16(os2.usWinAscent); // usWinAscent
    buf.writeInt16(os2.usWinDescent); // usWinDescent
    buf.writeInt32(os2.ulCodePageRange1); // ulCodePageRange1, Latin 1
    buf.writeInt32(os2.ulCodePageRange2); // ulCodePageRange2

    return buf;
}

module.exports = createOS2Table;
